solidity/contracts/isms/hook/layer-zero/LayerZeroV2Ism.sol

```diff
@@ -0,0 +1,134 @@
+
+    // @dev the offset of msg.data where the function parameters (as bytes) begins. 4 bytes is always used when encoding the function selector
+    uint8 constant FUNC_SELECTOR_OFFSET = 4;
+
+    // ============ Constructor ============
+    constructor(address _endpoint) {
+        require(
+            _endpoint != address(0),
+            "LayerZeroV2Ism: invalid authorized endpoint"
+        );
+        endpoint = _endpoint;
+    }
+
+    /**
+     * @notice Entry point for receiving msg/packet from the LayerZero endpoint.
+     * @param _message The payload of the received message.
+     * @dev Authorization verifcation is done within verifyMessageId() -> _isAuthorized()
+     */
+    function lzReceive(
+        Origin calldata,
+        bytes32,
+        bytes calldata _message,
+        address,
+        bytes calldata
+    ) external payable {
+        verifyMessageId(_messageId(_message));
+    }
+
+    // ============ Internal function ============
+
+    /**
+     * @notice Slices the messageId from the message delivered from LayerZeroV2Hook
+     * @dev message is created as abi.encodeCall(AbstractMessageIdAuthorizedIsm.verifyMessageId, id)
+     * @dev _message will be 36 bytes (4 bytes for function selector, and 32 bytes for messageId)
+     */
+    function _messageId(
+        bytes calldata _message
+    ) internal pure returns (bytes32) {
+        return bytes32(_message[FUNC_SELECTOR_OFFSET:]);
+    }
+
+    /**
+     * @notice Validates criterias to verify a message
+     * @dev this is called by AbstractMessageIdAuthorizedIsm.verifyMessageId
+     * @dev parses msg.value to get parameters from lzReceive()
+     */
+    function _isAuthorized() internal view override returns (bool) {
+        require(_isAuthorizedHook(), "LayerZeroV2Ism: hook is not authorized");
+
+        require(
+            _isVerifyMessageSelector(),
+            "LayerZeroV2Ism: message payload is incorrect"
+        );
+
+        require(
+            _isAuthorizedEndPoint(),
+            "LayerZeroV2Ism: endpoint is not authorized"
+        );
+
+        return true;
+    }
+
+    /**
+     * @notice check if origin.sender is the authorized hook
+     */
+    function _isAuthorizedHook() internal view returns (bool) {
+        (Origin memory origin, , , , ) = abi.decode(
+            msg.data[FUNC_SELECTOR_OFFSET:],
+            (Origin, bytes32, bytes, address, bytes)
+        );
+
+        return origin.sender == authorizedHook;
+    }
+
+    /**
+     * @notice check if the expected payload is the verifyMessageId selector, which is created in the authorized Hook
+     */
+    function _isVerifyMessageSelector() internal pure returns (bool) {
+        (, , bytes memory message, , ) = abi.decode(
+            msg.data[FUNC_SELECTOR_OFFSET:],
+            (Origin, bytes32, bytes, address, bytes)
+        );
+        return
+            bytes4(message) ==
+            AbstractMessageIdAuthorizedIsm.verifyMessageId.selector;
```

Comment:
instead of decoding `msg.data` why not just pass the `message` from `lzReceive`?

Reply:
As mentioned in chat
```
I wanted to consolidate all the validation data into _isAuthorized(), which is called by AbstractMessageIdAuthorizedIsm.verifyMessageId(). I can't pass anything into it since the signature is defined in AbstractMessageIdAuthorizedIsm, without any input.

The flow would be lzReceive() -> verifyMessageId() -> _isAuthorized() -> _isVerifyMessageSelector(). 

Again, the idea is to reuse AbstractMessageIdAuthorizedIsm as much as possible, but if you feel this isn't clear, i'm open to updates
```